src/main/java/org/jitsi/srtp/crypto/OpenSslAesGcmAuthOnlyCipherSpi.java

```diff
@@ -26,25 +27,44 @@
  * Can only be used for "decryption".
  */
 public class OpenSslAesGcmAuthOnlyCipherSpi
-    extends CipherSpi
+    extends CipherSpi implements AutoCloseable
 {
+    private static final class OpenSslAesGcmAuthOnlyCipherSpiCleanable implements Runnable
+    {
+        /**
+         * The context of the OpenSSL (Crypto) library through which the actual
+         * algorithm implementation is invoked by this instance.
+         */
+        long ptr;
+
+        @Override
+        public void run()
+        {
+            if (ptr != 0)
+            {
+                OpenSslAesGcmAuthOnlyCipherSpi.CRYPTO_gcm128_release(ptr);
+                ptr = 0;
+            }
+        }
+    }
+
     private static final int BLKLEN = 16;
 
-    private native long CRYPTO_gcm128_new();
+    private static native long CRYPTO_gcm128_new();
 
-    private native void CRYPTO_gcm128_release(long ctx);
+    private static native void CRYPTO_gcm128_release(long ctx);
 
-    private native boolean CRYPTO_gcm128_init(long ctx, byte[] key);
+    private static native boolean CRYPTO_gcm128_init(long ctx, byte[] key);
 
-    private native boolean CRYPTO_gcm128_setiv(long ctx, byte[] iv, int len);
+    private static native boolean CRYPTO_gcm128_setiv(long ctx, byte[] iv, int len);
 
-    private native boolean CRYPTO_gcm128_aad(long ctx,
+    private static native boolean CRYPTO_gcm128_aad(long ctx,
         byte[] in, int inOffset, int len);
 
-    private native boolean CRYPTO_gcm128_decrypt(long ctx,
+    private static native boolean CRYPTO_gcm128_decrypt(long ctx,
         byte[] out, int offset, int len);
 
-    private native boolean CRYPTO_gcm128_finish(long ctx,
+    private static native boolean CRYPTO_gcm128_finish(long ctx,
```

Comment:
> (Apparently stack frames don't count as references for the Java GC, if nothing in the stack frame is using the `this` pointer.). 

Uff, what kind of garbage is this!?

I'm actually surprised now this even compiled without me changing the corresponding C implementation from `jobject` to `jclazz`.
But this also means that I need to move the native method for cleaning into the inner classes.

Reply:
Well, in C a pointer-to-object is a pointer-to-object, and nothing in our native code dereferences those pointers...

Reply:
If I'm interpreting your comment correctly, ade2a3d30b0824d1ea37b19ffc3c15d36edc5b2e won't solve the GC issue.

Reply:
I think I changed all the methods that reference the ctx, other than the ones that actually do the free-ing, to be non-static -- since the free methods are only called from the Cleaner, that should be sufficient, even if not as clean as it could be. 

I agree that making the free'er a non-static method of the Cleaner would be, well, cleaner, but I think this should still work.

Is there a hole I'm missing?

Reply:
I had to read your original comment a few times more to get. You're right, it should work.

The signature of the native methods that are now static should still be changed for consistency (`jobject thiz` -> `jclass clazz`).

Reply:
Also, it'd be a good idea to comment in the Java code why the methods are not static.

Reply:
I'm reading up this and found `Finalization may occur whenever the virtual machine detects that no reference to an object will ever be stored in the heap: The garbage collector may reclaim an object even if the fields of that object are still in use, so long as the object has otherwise become unreachable.` in https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Reference.html#reachabilityFence-java.lang.Object-

I'm wondering if the non-static methods are really working (and I'm missing something from that sentence) or if that the non-static is just a hack that accidentally works most of the time. After all, in the scenario you described there isn't anything on the heap anymore - the stack doesn't count.

Reply:
The `jobject thiz` passed to the running JNI method counts a reference in the heap for these purposes, I’m pretty sure.  (I don’t see how it could be otherwise.)

Reply:
Somehow I'm not yet convinced reading the various SO posts. I'll see if I can trigger this behavior... unless you have a repro available when you found it out originally for abdf3cc24669c98b88329ea67f32573c95c92be1?

Reply:
I don't have a consistent repro for https://github.com/jitsi/jitsi-srtp/commit/abdf3cc24669c98b88329ea67f32573c95c92be1 unfortunately, but before my fix it happened pretty frequently in a JVB under load.  (Specifically it would happen in the final call to `deriveSessionKey` in `deriveSrtpKeys`.). You need a VM that's running close to its GC limit, of course.  Possibly you could see it by forcing a garbage collection in that method.

Reply:
It's not that I don't believe you it went away, this just really piqued my curiosity now :-)
Please go ahead with using the current state - I won't be able to do much anymore before next week.